File: src/TypeScript/src/finger-tree/ordering.ts
/** Total-order comparison function, following Array.sort's negative/zero/positive convention. */
export type Comparator<T> = (left: T, right: T) => number;

/** Natural JavaScript relational ordering for mutually comparable values. */
export function defaultComparator<T>(left: T, right: T): number {
    if (left === right) return 0;
    return left < right ? -1 : 1;
}

/** Returns a comparator with the opposite order. */
export function reverseComparator<T>(comparator: Comparator<T>): Comparator<T> {
    return (left: T, right: T): number => comparator(right, left);
}

File: src/Python/src/durable7/finger_tree/ordering.py
"""Ordering policies shared by ordered persistent collections."""

from __future__ import annotations

from collections.abc import Callable
from typing import TypeVar

T = TypeVar("T")
Comparator = Callable[[T, T], int]


def default_comparator(left: T, right: T) -> int:
    """Compare values using their native total ordering."""

    return -1 if left < right else 1 if left > right else 0  # type: ignore[operator]


def reverse_comparator(comparator: Comparator[T]) -> Comparator[T]:
    """Reverse an existing comparator while preserving zero."""

    def reversed_comparator(left: T, right: T) -> int:
        """Compare in the opposite order, leaving equivalence classes unchanged."""

        return comparator(right, left)

    return reversed_comparator


__all__ = ["Comparator", "default_comparator", "reverse_comparator"]
